src/Microsoft.Extensions.DependencyInjection.Abstractions/Extensions/ServiceCollectionDescriptorExtensions.cs

```diff
@@ -540,5 +540,38 @@ public static void TryAddSingleton<TService>(this IServiceCollection collection,
             collection.Add(descriptor);
             return collection;
         }
+
+        /// <summary>
+        /// Removes all services of type <typeparamef name="T"/> in <see cref="IServiceCollection"/>.
+        /// </summary>
+        /// <param name="collection">The <see cref="IServiceCollection"/>.</param>
+        /// <returns></returns>
+        public static IServiceCollection RemoveAll<T>(this IServiceCollection collection)
+        {
+            return RemoveAll(collection, typeof(T));
+        }
+
+        /// <summary>
+        /// Removes all services of type <paramef name="serviceType"/> in <see cref="IServiceCollection"/>.
+        /// </summary>
+        /// <param name="collection">The <see cref="IServiceCollection"/>.</param>
+        /// <param name="serviceType">The service type to remove.</param>
+        /// <returns></returns>
+        public static IServiceCollection RemoveAll(this IServiceCollection collection, Type serviceType)
+        {
+            if (serviceType == null)
+            {
+                throw new ArgumentNullException(nameof(serviceType));
+            }
+
+            foreach (var serviceDescriptor in collection
```

Comment:
Indexes change when you start removing things, or it would need to go in reverse order. I found this one simplest to understand.